java/arcs/core/host/ArcHostContextParticle.kt

```diff
@@ -0,0 +1,255 @@
+                    it.consecutiveFailures.toInt()
+                )
+            }.associateBy({ it.first }, { it.second })
+
+            return ArcHostContext(particles.toMutableMap(), arcStateEntity?.let {
+                ArcState.valueOf(it.arcState)
+            } ?: ArcState.NeverStarted)
+        } catch (e: Exception) {
+            throw IllegalStateException("Unable to deserialize $arcId for $hostId", e)
+        }
+    }
+
+    private fun createHandlesMap(
+        handleConnections: Map<String, List<Pair<String, Pair<String, Plan.HandleConnection>>>>,
+        it: ArcHostParticle_Particles,
+        arcId: String
+    ): Map<String, Plan.HandleConnection> {
+        return handleConnections[it.particleName]?.associateBy(
+            { it.second.first }, // handleName
+            { it.second.second } // Plan.HandleConnection
+        ) ?: throw IllegalArgumentException(
+            "Can't find handleConnection for ${it.particleName} in $arcId"
+        )
+    }
+
+    private fun createHandleConnection(
+        entity: ArcHostParticle_HandleConnections,
+        arcId: String,
+        hostId: String,
+        instantiatedParticles: Map<String, Particle>
+    ): Pair<String, Plan.HandleConnection> {
+        return (entity.handleName to Plan.HandleConnection(
+            StorageKeyParser.parse(entity.storageKey),
+            HandleMode.valueOf(entity.mode),
+            fromTag(
+                arcId,
+                requireNotNull(instantiatedParticles[entity.particleName]) {
+                    "${entity.particleName} not instantiable for $arcId and $hostId"
+                },
+                entity.type,
+                entity.handleName
+            ),
+            entity.ttl.let { num -> if (num != 0.0) Ttl.Minutes(num.toInt()) else null }
+        ))
+    }
+
+    /**
+     * Using instantiated particle to obtain [Schema] objects throught their
+     * associated [EntitySpec], reconstruct an associated [Type] object.
+     */
+    fun fromTag(arcId: String, particle: Particle, tag: String, handleName: String): Type {
+        try {
+            val schema = particle.handles.getEntitySpec(handleName).schema()
+            return when (Tag.valueOf(tag)) {
+                Tag.Singleton -> SingletonType(EntityType(schema))
+                Tag.Collection -> CollectionType(EntityType(schema))
+                Tag.Entity -> EntityType(schema)
+                else -> throw IllegalArgumentException(
+                    "Illegal Tag $tag when deserializing ArcHostContext"
+                )
+            }
+        } catch (nse: NoSuchElementException) {
+            throw NoSuchElementException("""
+                Can't create Type $tag for Handle $handleName and ${particle::class}. This usually
+                occurs because the Particle or ArcHost implementation has changed since
+                the last time this arc was serialized. ${nse.message}
+            """.trimIndent())
```

Comment:
Also, supply the original exception as the cause for this one.

Reply:
Changed to IllegalStateException, because NoSuchElementException doesn't support supplying the cause.